### PR TITLE
Add types to reducers

### PR DESCRIPTION
**Pull request by sfentress**
This is a RFC as much as a PR as I'm not certain this is the best solution for typing Redux, but it seems to work generally well so far, with not too much boilerplate (but some redundancy) added to the reducers.

This adds types to all the models used in the reducers. These types support Immutable's Maps (mostly turned into Records here), including the nested setters and getters:

```
state.getIn(["report", "clazzName"]);       // <- type string
state.setIn(["report", "clazzName"], 2);    // <- error
```

This required switching the top-level `Maps` to `StaticallyTypedRecords` lifted with modifications from [this blog post](https://blog.mgechev.com/2018/01/18/react-typescript-redux-immutable/). The only major downside is the ugliness/repetition introduced in this commit: 422f28aa. We need to repeat all the properties of the state a second time when falsely-claiming that we implement an interface, so that the `getIn` and `setIn` type definitions can go down the chain to find the types. This also required silencing a Typescript rule so as not to doubly-set values for these props.

**To do, now or later**: There are probably further places in the code that we could be using `RootState` that I missed, these could be included in this PR if found or added as we keep converting stuff. Also, most of the `Maps` are typed as `<any, any>` to simplify this work without getting bogged down into tracking how the maps are used everywhere. These should get typed when code is written that uses them, however.

@pjanik I included you as a reviewer since you've done a lot of the Redux work, but feel free to ignore if you're busy elsewhere.

Also note this is currently based on #153 so as to simplify the later rebase, as that also modifies the reducers.

### CODE REVIEW

File: js/reducers/view-reducer.ts [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+    // case SWITCH_REPORT_VIEW:
+    //   return state.set('type', action.viewType)
+    default:
+      return state;
+  }
+}
+
+export function isDashboardView(viewType: string) {
+  return (viewType === DASHBOARD) || (viewType === PORTAL_DASHBOARD);
+}
```
Comment (kswenson) [inline js/reducers/view-reducer.ts:44]:
```suggestion
}

```
EOL

File: js/reducers/view-reducer.ts [hunk 2]
```diff
@@ -0,0 +1,44 @@
...
+
+export class ViewState extends INITIAL_VIEW_STATE implements IViewState {
+  constructor(config: Partial<IViewState>) {
+    super(config);
+  }
+  type: string;
+}
+
+export default function view(state = new ViewState({}), action: any) {
+  switch (action && action.type) {
```
Comment (kswenson) [inline js/reducers/view-reducer.ts:32]:
```suggestion
  switch (action?.type) {
```

File: js/reducers/data-reducer.ts [hunk 3]
```diff
@@ -0,0 +1,37 @@
...
+      return state.set("isFetching", true);
+    case RECEIVE_RESOURCE_STRUCTURE:
+      return state.set("isFetching", false);
+    case FETCH_ERROR:
+      return state.set("isFetching", false)
+                  .set("error", action.response);
+    default:
+      return state;
+  }
+}
```
Comment (kswenson) [inline js/reducers/data-reducer.ts:37]:
```suggestion
}

```
EOL

File: js/reducers/report-reducer.ts [hunk 4]
```diff
@@ -163,15 +115,12 @@ function report(state = INITIAL_REPORT_STATE, action) {
   // from JWT info) and Firestore security rules also ensure that.
   const { studentId } = queryString.parse(window.location.search);
   const urlBasedStudentSelection = studentId ? Immutable.fromJS([ studentId ]) : null;
-  switch (action.type) {
+  switch (action && action.type) {
```
[the remark this answers is not in the document]
Comment (sfentress) [reply js/reducers/report-reducer.ts]: Oops, accidentally left-over from before I was initializing these with a `dummyAction`. These can all just be `switch (action.type)`

File: js/reducers/report-reducer.ts [hunk 5]
```diff
@@ -248,12 +197,54 @@ function report(state = INITIAL_REPORT_STATE, action) {
...
+    });
+    return state;
   });
 }
+
+function setUserSettings(state: ReportState, response: any) {
+  const {visibility_filter, anonymous_report} = response;
+  let selectedQuestions: any[] = [];
+  // The visibility_filter might not be setup on the user settings yet
+  if (visibility_filter && visibility_filter.questions) {
```
[the remark this answers is not in the document]
Comment (sfentress) [reply js/reducers/report-reducer.ts]: :+1:

File: js/reducers/report-reducer.ts [hunk 6]
```diff
@@ -248,12 +197,54 @@ function report(state = INITIAL_REPORT_STATE, action) {
...
+    return state;
+  });
+}
+
+function setAnonymous(state: ReportState, anonymous: boolean) {
+  let idx = 1;
+  const newStudents = state.get("students")
+    .map(s => s.set("name", anonymous ? `Student ${idx++}` : s.get("realName"))) as Map<any, any>;
+  return state.set("anonymous", anonymous).set("students", newStudents);
+}
```
[the remark this answers is not in the document]
Comment (sfentress) [reply js/reducers/report-reducer.ts]: :+1:

**Pull request merged**